### ISSUE TITLE
catchError not working as intended (docs)

### ISSUE THREAD

**Issue by cmarques2** [dart-lang/language#2159]
### From the documentation

```
myFunc().then((value) {
  doSomethingWith(value);
  ...
  throw Exception('Some arbitrary error');
}).catchError(handleError);
```

If myFunc()’s Future completes with a value, then()’s callback fires. **If code within then()’s callback throws** (as it does in the example above), then()’s Future completes with an error. **That error is handled by catchError()**.

That doesn't appear to be working.

[Dartpad Example](https://dartpad.dev/?id=916f447423dcd28b490d5457dacc2ab7)

**Comment by mateusfccp** [comment]
Considering the provided DartPad example, if I understood correctly what you are saying, here is what's happening:

* In the `.then` body of the provided example, you only provide a `throw`, which makes your future a `Future<Never>`;
* The `catchError`, below, expects the return type to be the same as the `.then`, but it returns a String;
* By providing `<String?>` to your `.then` or making your `catchError` return a `Never` makes it work.

**Comment by cmarques2** [comment]
Thank you, that was very helpful in figuring out that example. I wish I could update the docs with this. :)

**Comment by mateusfccp** [comment]
Yeah, I don't know which documentation you are referring to, but there were many changes in the type system since null-safety, maybe some of them is outdated...

If you feel like, you may open an issue in the documentation repository and provide a PR.
